### PR TITLE
feat: recursively walk symlinks pointing at dirs

### PR DESCRIPTION
**Pull request by booxter**
Before:

```
/tmp/test-recursive-dir-symlinks
├── dir
│  └── test.txt
└── symlinkdir -> /tmp/test-recursive-dir-symlinks/dir
```

After:

```
/tmp/test-recursive-dir-symlinks
├── dir
│  └── test.txt
└── symlinkdir -> /tmp/test-recursive-dir-symlinks/dir
   └── test.txt
```

Signed-off-by: Ihar Hrachyshka <ihar.hrachyshka@gmail.com>

##### Description

This will make `eza` resolve symlinks pointing at valid directories like a
regular directories, meaning: if requested, unravel them recursively. This is
handy in cases where an inspected directory contains multiple layers of symlink
indirection (hi nix!)

I've confirmed that loops (situations where symlinks from two directories
cross-point to each other) are handled gracefully by `eza` already (reporting
an error but proceeding to handle the rest of the files.) See below.

```
...
│  ├── links -> /tmp/test/test/test
│  │  └── link4 -> /tmp/test/test2
│  │     └── link3 -> /tmp/test/test/test
│  │        └── link4 -> /tmp/test/test2
│  │           └── link3 -> /tmp/test/test/test
│  │              └── link4 -> /tmp/test/test2
│  │                 └── link3 -> /tmp/test/test/test
│  │                    └── link4 -> /tmp/test/test2
│  │                       └── link3 -> /tmp/test/test/test
│  │                          └── link4 -> /tmp/test/test2
│  │                             └── link3 -> /tmp/test/test/test
│  │                                └── link4 -> /tmp/test/test2
│  │                                   └── link3 -> /tmp/test/test/test
│  │                                      └── link4 -> /tmp/test/test2
│  │                                         └── link3 -> /tmp/test/test/test
│  │                                            └── link4 -> /tmp/test/test2
│  │                                               └── <Too many levels of symbolic links (os error 62)>
│  └── test
│     └── link4 -> /tmp/test/test2
│        └── link3 -> /tmp/test/test/test
...
```

##### How Has This Been Tested?

Local test env: nix-darwin. Used included nix targets to validate the patch.

Confirmed that the test suite passed. (Only failures specific to nix-darwin are
observed.)

Checked on a a number of synthetic and real nested directories with several
layers of symlink indirection.

Checked that loops are properly handled by the tool.

---

This patch changes behavior of the tool. I am not sure this change justifies a
new CLI flag though. The first version of the patch enables it unconditionally.

### CODE REVIEW

**Review by cafkafk — DISMISSED** [review]
This is cool, and looks like it works without issues, thanks!

> This patch changes behavior of the tool. I am not sure this change justifies a
new CLI flag though. The first version of the patch enables it unconditionally.

I do think I'm gonna spend a brief amount of time just making sure how (a probably unrepresentative) sample of people feel about making this the new default... We may want to have a flag for disabling it, but that doesn't have to be this PR.

I do like the idea of this being the default behavior. I'll merge after making sure other people seem to agree.

**Comment by daviessm** [comment]
> I do like the idea of this being the default behavior. I'll merge after making sure other people seem to agree.

I'm not sure about this. Is it only for tree mode? I can definitely see it being used behind a flag/alias but I wouldn't want to have to disable it each time.

**Comment by booxter** [comment]
@daviessm non-tree version works the same way. The included test changes confirm this. (Side note: I am wondering if perhaps the suite should have another set of tests for tree mode? I haven't seen any.)

**Comment by booxter** [comment]
About the (default) behavior, a data point: `tree` does what `eza` is doing right now (**not** walking down the symlink rabbit hole).

It shouldn't be hard to make it an opt-in controlled by a flag (that people can then put in their aliases - I would.) I don't see a reason to disable this proposed behavior myself, but I may miss how other people use the tool.

**Comment by cafkafk** [comment]
> It shouldn't be hard to make it an opt-in controlled by a flag (that people can then put in their aliases - I would.) I don't see a reason to disable this proposed behavior myself, but I may miss how other people use the tool.

I've asked a few people, and most seem to be fine with it being default **if** there is cycle detection, which is probably gonna be pretty compute expensive *or* require using a lot of platform specific code, e.g. `man fts` functions on Linux/BSD/MacOS (which also isn't in musl without external libraries), and some pretty obscure windows ones.

I think the best thing would be to start with just the scope of this PR, behind a flag, and then if people want to start tackling cycle detection (and manage to do so in a way that doesn't grind eza to a halt), we can make it the default behavior

**Comment by booxter** [comment]
@cafkafk I can add an option. Before I do, I'd like to ask what the reason for force-pushing to the branch is. I don't see any visible changes.

---

About symbolic link loop detection. On POSIX systems, `opendir` already returns `ELOOP` on a loop detection. That's the error 62 from the snippet in the description. So I don't think Unix would need anything special.

As for Windows, I don't know if `FindFirstFile` that, apparently, `std::fs::read_dir` uses implements a similar behavior. We can try and see (I don't have access to win32 machine at the moment though.)

File: tests/gen/long_time_style_custom_non_recent_none_nix.stderr
Comment (booxter) [inline tests/gen/long_time_style_custom_non_recent_none_nix.stderr:1]: these checks are kinda stupid. the tests shouldn't assume a particular line of error. not sure if it would be easy to make them more robust though.

**Comment by booxter** [comment]
@cafkafk I've added a flag to control the new behavior. Let me know if it's good enough now.

File: tests/gen/long_time_style_custom_non_recent_none_nix.stderr
Comment (MartinFillon) [reply tests/gen/long_time_style_custom_non_recent_none_nix.stderr:1]: These tests are getting more robust in #787 which is a whole refactor of the args parsing

**Review by MartinFillon — DISMISSED** [review]
lgtm, thanks for the pr and idea, this will be really useful for some edge cases as you mentionned

**Comment by cafkafk** [comment]
> Before I do, I'd like to ask what the reason for force-pushing to the branch is. I don't see any visible changes.

I think this was just me doing a `git pull origin main --rebase` on the branch, so no changes to PR content :p

**Comment by cafkafk** [comment]
I've resolved the conflict with the annoying tests here, I'll review the code and potentially regenerate

**Review by cafkafk — CHANGES_REQUESTED** [review]
Okay, I've looked through the code now, and done reasonable testing for performance regressions and behavior. It's solid and ready to merge. I did spot that this is lacking completions however.

All this needs before merge is adding the new option to the completions files.

Thanks for making this, it's very useful!

**Comment by booxter** [comment]
@cafkafk I've updated completion files. I haven't tested this though (not sure how to do it with nix env, nor I know how nushell and fish even work). I also haven't touched bash completion since it looks like it extracts long options from `--help`. Let me know if this is ok.

**Review by cafkafk — APPROVED** [review]
Okay, thanks, then it's done! I'll merge now, thanks for the contribution! This is very useful for someone with a lot of symlink farms... like me :D

---

> not sure how to do it with nix env

One possible way is to `nix profile install .` in the root of the repository, and then opening a new terminal, that should work *in most cases*.

**Pull request merged**
